src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java

```diff
@@ -250,15 +250,15 @@ public void handleNotification(INotification notification, Object sender)
             SSTableAddedNotification notice = (SSTableAddedNotification) notification;
 
             // Avoid validation for index files just written following Memtable flush.
-            boolean validate = !notice.memtable().isPresent();
+            IndexValidation validate = notice.memtable().isPresent() ? IndexValidation.NONE : IndexValidation.CHECKSUM;
```

Comment:
So this is the crux of the change, correct? (i.e. SSTables added via streaming or import will be checksummed, unlike before, where they would undergo only header/footer validation?)

Reply:
One of the general things I looked at while reviewing here was whether the notification order of the different `INotificationConsumer` things matters. `StorageAttachedIndexGroup` and `SecondaryIndexManager` in particular are the interesting ones. Basically...do we checksum the things we should and not the things we shouldn't, given we happen to have SIM registered before the index group? Here's what I've come up with:

1.) When we stream entire SSTables w/ their attached index components (or locally import new SSTables w/ index components already built), SIM handles the notification, but the index build it triggers should be a no-op. Even if it built something it wouldn't register as a full rebuild and no checksumming would be done. Then when `StorageAttachedIndexGroup` handles the notification here, it does the checksum, effectively protecting us from corruption introduced by a streaming source. This all seems reasonable.

2.) When we don't stream an entire SSTable during repair or partial SSTable bootstrapping (or locally import new SSTables w/ index components not already built), SIM handles the notification and blocks to build SSTable-attached index bits. Then (again, this is an accident of our ordering) `StorageAttachedIndexGroup` handles the notification. It sees that there isn't a Memtable and validates the checksum. _This isn't the end of the world, but it would seem unnecessary, given we just build the indexes from the streamed SSTables._ If we knew that all of the added SSTables didn't come with SAI components, we'd bypass checksumming, right?

I'm not sure how exactly we want to address this, or whether we would want to address it in this patch (although that might be nice, since we have it loaded into our working memories), but perhaps we could agree that it's an issue (or not, and that I'm missing something) and go from there?

The two things that come to mind to "fix" this are a.) having more information in the notification or b.) having the index group handle notifications *before* SIM does. (SIM only handles `SSTableAddedNotification` FWIW).

CC @mike-tr-adamson @adelapena

Reply:
We have been looking at this in some detail today to try and establish the best way to correct it.

From our perspective, the `SSTableAddedNotification` is handled in 2 places. It is handled by the `SecondaryIndexManager` that with invoke a `fullRebuild = false` index build on the added SSTables, and it is handled by the `StorageAttachedIndexGroup` that will validate and register the SSTables. 

The problem with the current implementation is:
- The invoked index build will rebuild the per-column index components for `fullRebuild = false` even if the components are present and valid. We do, though, only build the per-SSTable components if they are invalid, in this case.
- The SAIG handler will validate the components even though they have been already validated by the build.

What should be noted here is that the SIM will only invoke an index build if a memtable is not present in the notification, and the SAIG currently only validates the index components if the memtable isn't present but still registers the components in either case.

We feel that the following changes could be made to improve this:
1. The index build should only build the per-column indexes if `fullRebuild=true`, the index doesn't exist, or the index is invalid. In either case, the SSTables must be registered (on build completion or immediately).
2. The SAIG should only register the SSTables if the memtable is present and shouldn't validate the components because this is the result of a memtable flush.

This will require some rework of the index builder, to make sure that the SSTables are always registered correctly and, as such, we would prefer that this was done on a separate ticket from this one. Which phase that ticket goes into is up for discussion. The downsides of the current implementation are that we can rebuild indexes that already have valid components and we can end up validating components twice. 

WDYT?

Reply:
@mike-tr-adamson @pkolaczk I'm okay w/ doing this in another Jira. There's enough risk to make sure we focus on it specifically, and after this patch we'll only be doing something sub-optimally, not incorrectly. Now...in solution space...

> The index build should only build the per-column indexes if fullRebuild=true, the index doesn't exist, or the index is invalid. In either case, the SSTables must be registered (on build completion or immediately).

I don't see any problem w/ this, although I'm not sure I understand the problem in the streaming context.

> The invoked index build will rebuild the per-column index components for fullRebuild = false even if the components are present and valid. We do, though, only build the per-SSTable components if they are invalid, in this case.

Right after streaming, no index components for the new SSTables are present, right?

> The SAIG should only register the SSTables if the memtable is present and shouldn't validate the components because this is the result of a memtable flush.

`StorageAttachedIndexGroup` registering and validating newly streamed entire index components is what makes this patch work, right? If we stopped doing that, who would register/checksum when the `SIM`-triggered build is a no-op?

Reply:
See [CASSANDRA-18653](https://issues.apache.org/jira/browse/CASSANDRA-18653)